### PR TITLE
error to warning

### PR DESCRIPTION
**Pull request by tom-wa**

### CODE REVIEW

**Comment by markus2330** [comment]
The `testscr_check_formatting` unittest finally works ;)

I think the resetting of an error between `kdbGet`/`kdbSet` calls is still missing. (And restoring of old errors if no new error happened).

**Comment by tom-wa** [comment]
it's currently just turning every error into a warning after the first error is set. 
about the `kdbGet` / `kdbSet` error resetting part, i'm not really sure what you mean

**Comment by markus2330** [comment]
Let us discuss this today. It should implement two rules:

1. *Within* an individual `kdbOpen`, `kdbGet` or `kdbSet`-calls only the first error will be used
2. *Between* individual `kdbOpen`, `kdbGet` or `kdbSet`-calls the error is only set iff the call is non-successful  (those returning -1)

Rule 2 means that that a new call with a new error overwrites the previous error *and also means* a new successful call keeps the previous error. (errno semantics)

This should also be documented in the API and kdb testcases should be added.

**Comment by markus2330** [comment]
jenkins build please

**Comment by markus2330** [comment]
any update here?

**Comment by markus2330** [comment]
jenkins build please

**Comment by markus2330** [comment]
jenkins build fast please

**Comment by markus2330** [comment]
jenkins build fast please

**Comment by markus2330** [comment]
jenkins build please

**Comment by markus2330** [comment]
jenkins build stable please

**Comment by markus2330** [comment]
jenkins build unstable please

**Comment by markus2330** [comment]
any update here?

**Comment by tom-wa** [comment]
I'm having a few issues getting the second part to work

**Comment by markus2330** [comment]
Which are? You simply write guards at beginning/end of `kdb*`-functions and make sure that errors are only set according the API description and transform them to warnings otherwise.

**Comment by tom-wa** [comment]
@markus2330 can you take a look at kdbSet here ? 
it basically just backs up the error, removes it if a new error could be set and -1 returned and restores it otherwise.

**Comment by markus2330** [comment]
I did not look into the details, but yes it is something like this.

Can you add test cases for different errors scenarios in tests/kdb/testkdb_simple.cpp
(Btw. you can use the C API too, you only need `using namespace ckdb;`)
It needs to be in this file, because the text fixture Simple makes sure that a new mountpoint exists in every testcase for `testRoot`.

**Comment by markus2330** [comment]
We need extensive tests here, let us do it for next release.

**Comment by markus2330** [comment]
@mpranj Is it okay if we merge the PR for this release?

File: src/libs/elektra/kdb.c [hunk 1]
```diff
@@ -633,6 +633,34 @@ static int elektraGetDoUpdateWithGlobalHooks (KDB * handle, Split * split, KeySe
 	}
 	return 0;
 }
+
+static int copyError (Key * dest, Key * src)
+{
+	keyRewindMeta (src);
+	Key * metaKey = (Key *)keyGetMeta (src, "error");
```
Comment (markus2330) [inline src/libs/elektra/kdb.c]: avoid cast by using `const Key`

**Comment by mpranj** [comment]
@markus2330 you mean this PR, or were you hinting at the hooks?
This one here can be merged!

**Comment by markus2330** [comment]
Yes, I mean this PR.

@tom-wa Can you rebase, fix the small issue, and add some test cases?

**Comment by tom-wa** [comment]
isn't that error actually expected on the second call to kdbSet ?
resolver (with elektraOpenFile) gets called before the error plugin and because kdbSet fails the new error should win

**Comment by markus2330** [comment]
> isn't that error actually expected on the second call to kdbSet

No, from user perspective the last error should win (like working with errno). Only from plugin perspective the first error should win (and others should be attached as warning). So you need to clean the error in the beginning and restore the error on success (like errno is not touched on success).

Please also add some more test cases, it would be great to have this for 0.8.19.

**Comment by markus2330** [comment]
Thanks, you are right about error 30, we have to investigate it. Seems we had a silent failure.

@tom-wa But there are still test cases missing. Could you please add them?

**Comment by markus2330** [comment]
Thanks, your improved error handling actually led to a fix 4637e1f5fa9a410aeb65928ee5b74ff273020d94 of a quite serious issue (conflict of later kdbSet() against non-missing file if earlier kdbSet() had no file and is aborted).

Can you add some more test cases here? Maybe we find another issue ;)

**Comment by markus2330** [comment]
jenkins build mergerequests please

**Comment by markus2330** [comment]
jenkins build please

**Comment by markus2330** [comment]
jenkins build debug please

**Comment by markus2330** [comment]
Thank you!

**Pull request merged**